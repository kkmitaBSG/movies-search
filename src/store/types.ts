export const api_key = "92b418e837b833be308bbfb1fb2aca1e";

export enum MoviesActionType {
  getMovies = "GET_MOVIES",
  getMoviesFail = "GET_MOVIES_FAIL",
  getSearchMovies = "GET_SEARCH_MOVIES",
  getMoviesNextPage = "GET_MOVIES_NEXT_PAGE",
  getMovieDetails = "GET_MOVIE_DETAILS",
}

export interface Movie {
  id: string;
  title: string;
  overview: string;
  poster_path?: string;
  poster?: string;
}

export interface Movies {
  results?: Movie[];
  page: number;
}

export interface MovieDetails {
  id?: string;
  backdrop_path?: string;
  title?: string;
  backdrop?: string;
  overview?: string;
  production_countries?: Array<{ name: string }>;
  release_date?: string;
  genres?: Array<{ name: string }>;
}

export interface MovieDetailsState {
  [id: string]: MovieDetails;
}

export interface State {
  movies: Movies;
  movieDetails: MovieDetails;
}

export const moviePosterOriginalPath = "https://image.tmdb.org/t/p/w500";
