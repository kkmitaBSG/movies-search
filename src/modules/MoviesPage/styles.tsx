import { NavLink } from "react-router-dom";
import styled from "styled-components";

export const Title = styled(NavLink)`
  display: inline-block;
  font-size: 40px;
  text-decoration: none;
  font-weight: bold;
  color: #fff;
  margin: 30px 0;
`;

export const Container = styled.div`
  margin: 0 60px;
`;

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Loader = styled.div`
  width: 100%;
  margin: 50px;
  text-align: center;
`;
