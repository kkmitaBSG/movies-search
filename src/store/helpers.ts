import { Movie, MovieDetails, moviePosterOriginalPath } from "./types";

export const transformMovieData = (movies?: Movie[]) =>
  movies?.map((movie) => ({
    ...movie,
    poster: moviePosterOriginalPath + movie.poster_path,
  }));

export const transformMovieDetailsData = (movieDetail?: MovieDetails) => ({
  ...movieDetail,
  backdrop: moviePosterOriginalPath + movieDetail?.backdrop_path,
});
