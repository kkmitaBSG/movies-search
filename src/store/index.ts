import { configureStore } from "@reduxjs/toolkit";
import moviesReducer from "./reducers/moviesReducer";
import movieDetailsReducer from "./reducers/movieDetailsReducer";
import createSagaMiddleware from "redux-saga";
import rootSaga from "./sagas";

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    movies: moviesReducer,
    movieDetails: movieDetailsReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
