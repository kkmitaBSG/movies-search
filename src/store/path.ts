const API_PATH = `https://api.themoviedb.org/3`;

const MOVIE_PATH = `${API_PATH}/movie`;

export const MOVIE_POPULAR_PATH = `${MOVIE_PATH}/popular`;

export const MOVIE_DETAILS_PATH = `${MOVIE_PATH}/[id]`;

export const SEARCH_MOVIE_PATH = `${API_PATH}/search/movie`;
