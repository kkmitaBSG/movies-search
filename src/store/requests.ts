import axios from "axios";
import { api_key, Movie, Movies } from "./types";
import {
  MOVIE_DETAILS_PATH,
  MOVIE_POPULAR_PATH,
  SEARCH_MOVIE_PATH,
} from "./path";

export const getMovies = (pageNum: number) =>
  axios.get<Movies[]>(
    MOVIE_POPULAR_PATH + `?api_key=${api_key}&page=${pageNum}`
  );

export const getSearchMovies = (searchPhrase: string, pageNum: number) =>
  axios.get<Movies[]>(
    SEARCH_MOVIE_PATH +
      `?api_key=${api_key}&page=${pageNum}&query=${searchPhrase}`
  );

export const getMovieDetails = (movieId: string) =>
  axios.get<Movie>(
    MOVIE_DETAILS_PATH.replace("[id]", movieId) + `?api_key=${api_key}`
  );
