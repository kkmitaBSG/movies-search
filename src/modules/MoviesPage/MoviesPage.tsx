import { useCallback, useEffect } from "react";
import throttle from "lodash.throttle";
import messages from "../../constants/messages";
import Item from "../../components/Item/Item";
import { useDispatch, useSelector } from "react-redux";
import { MoviesActionType } from "../../store/types";
import { getMovies } from "../../store/selectors";
import { useSearchParams } from "react-router-dom";
import SearchBar from "../../components/SearchBar/SearchBar";
import * as S from "./styles";

const MoviesPage = () => {
  const dispatch = useDispatch();
  const [searchParams] = useSearchParams();
  const { results } = useSelector(getMovies);

  useEffect(() => {
    const searchPhrase = searchParams.get("search");
    dispatch({
      type: MoviesActionType.getMovies,
      searchPhrase,
    });
  }, [dispatch, searchParams]);

  const handleScroll = throttle(() => {
    const searchPhrase = searchParams.get("search");
    if (
      window.innerHeight + window.scrollY >=
      document.body.offsetHeight - 20
    ) {
      dispatch({ type: MoviesActionType.getMoviesNextPage, searchPhrase });
    }
  }, 300);

  useEffect(() => {
    window.addEventListener("scroll", handleScroll);

    return () => {
      window.removeEventListener("scroll", handleScroll);
    };
  }, [dispatch, handleScroll]);

  const onInputChange = useCallback((searchPhrase: string) => {
    dispatch({ type: MoviesActionType.getMovies, searchPhrase });
  }, [dispatch]);

  const items = results?.map((item) => <Item key={item.id} {...item} />);

  return (
    <S.Container>
      <S.Title to="/">{messages.moviesTitle}</S.Title>
      <SearchBar onInputChange={onInputChange} />

      <S.Wrapper>{items}</S.Wrapper>
    </S.Container>
  );
};

export default MoviesPage;
