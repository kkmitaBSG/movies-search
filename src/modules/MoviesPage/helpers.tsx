import throttle from "lodash.throttle";
import { AnyAction, Dispatch } from "redux";
import { MoviesActionType } from "../../store/types";

export const handleScroll = throttle(
  (searchParams: URLSearchParams, dispatch: Dispatch<AnyAction>) => {
    const searchPhrase = searchParams.get("search");
    if (
      window.innerHeight + window.scrollY >=
      document.body.offsetHeight - 400
    ) {
      dispatch({ type: MoviesActionType.getMoviesNextPage, searchPhrase });
    }
  },
  300
);
