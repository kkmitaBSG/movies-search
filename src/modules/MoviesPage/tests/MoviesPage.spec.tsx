import * as reactRedux from "react-redux";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import MoviesPage from "../MoviesPage";
import { Movies } from "../../../store/types";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
}));

jest.mock("../../../components/Item/Item");

const mockData: Movies = {
  results: [
    {
      id: "1",
      title: "item1Name",
      overview: "overview1",
    },
    {
      id: "2",
      title: "item2Name",
      overview: "overview2",
    },
    {
      id: "3",
      title: "item3Name",
      overview: "overview3",
    },
  ],
  page: 1,
};

describe("MoviesPage", () => {
  const useSelectorMock = jest.spyOn(reactRedux, "useSelector");
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();
  });

  test("render with movies", () => {
    useSelectorMock.mockReturnValue(mockData);

    const component = renderer.create(
      <MemoryRouter>
        <MoviesPage />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
