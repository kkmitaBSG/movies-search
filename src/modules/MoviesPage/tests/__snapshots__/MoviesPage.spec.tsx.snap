// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesPage render with movies 1`] = `
<div
  className="sc-dkrFOg gsGYiQ"
>
  <a
    aria-current="page"
    className="sc-gswNZR dTKyTL active"
    href="/"
    onClick={[Function]}
  >
    Movies
  </a>
  <input
    className="sc-bcXHqe cEoBZN"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search a movie..."
    type="text"
    value=""
  />
  <div
    className="sc-hLBbgP eQAeNk"
  >
    ItemMockedComponent
    ItemMockedComponent
    ItemMockedComponent
  </div>
</div>
`;
